[ABLD-28] Add macOS build for AArch64/ARM64

rdesgroppes opened the pull request:
### What does this PR do?
This is the payload for adding macOS "native" build support for the AArch64/ARM64 architecture to the Agent.

### Motivation
Remove the need for Rosetta 2 translation at runtime.

### Describe how you validated your changes
x86_64/AMD64 macOS jobs get here duplicated for AArch64/ARM64, bringing the same automated coverage to both architectures.

Colleagues running macOS Silicon are expected to try out the newly built Agent locally.

### Possible Drawbacks / Trade-offs
Some follow-ups:
- [x] DataDog/datadog-agent-dev#137
- [x] DataDog/datadog-agent-buildimages#888
- [x] DataDog/datadog-agent#38446
- [x] DataDog/datadog-agent#39241
- [x] DataDog/datadog-agent#39260
- [x] DataDog/datadog-agent#39641
- [x] DataDog/omnibus-ruby#241
- [x] DataDog/omnibus-ruby#242
- [x] DataDog/omnibus-ruby#243
- [x] DataDog/integrations-core#20812
- [x] DataDog/integrations-core#20815
- [ ] DataDog/integrations-core#20645
- [x] DataDog/ci-platform-machine-images#343
- [x] DataDog/ci-platform-machine-images#391
- [x] DataDog/ci-platform-machine-images#392
- [x] DataDog/agent-release-management#365
- [x] DataDog/ansible-datadog#664

### Additional Notes
Groundwork enabling the present payload:
- [x] DataDog/integrations-core#20455
- [x] DataDog/integrations-core#20577
- [x] DataDog/integrations-core#20699
- [x] DataDog/omnibus-ruby#239
- [x] DataDog/omnibus-ruby#240
- [x] DataDog/datadog-agent#38498
- [x] DataDog/datadog-agent#39082
- [x] DataDog/agent-release-management#364

Some earlier scouting:
- DataDog/integrations-core#19193 (credits to @Pythyu)
- DataDog/omnibus-software#553 (credits to @Pythyu)
- DataDog/datadog-agent#18173 (credits to @timsutton)
- DataDog/datadog-agent#31787 (credits to @Pythyu)
- DataDog/ci-platform-machine-images#321

rdesgroppes committed:
Add macOS build for AArch64/ARM64

This is the payload for adding macOS "native" build support for the
AArch64/ARM64 architecture to the Agent, that is by removing the need
for Rosetta 2 translation at runtime.

Groundwork enabling the present payload:
- DataDog/integrations-core#20455
- DataDog/integrations-core#20577
- DataDog/integrations-core#20699
- DataDog/omnibus-ruby#239
- DataDog/omnibus-ruby#240
- DataDog/datadog-agent#38498
- DataDog/datadog-agent#39082
- DataDog/agent-release-management#364

Some earlier scouting:
- DataDog/integrations-core#19193 (credits to @Pythyu)
- DataDog/omnibus-software#553 (credits to @Pythyu)
- DataDog/datadog-agent#31787 (credits to @Pythyu)
- DataDog/ci-platform-machine-images#321

Some follow-ups:
- DataDog/datadog-agent-dev#137
- DataDog/datadog-agent-buildimages#888
- DataDog/datadog-agent#38446
- DataDog/integrations-core#20812
- DataDog/integrations-core#20815
- DataDog/integrations-core#20645
- Datadog/ci-platform-machine-images#343
- DataDog/agent-release-management#365

rdesgroppes commented on release.json:
```diff
@@ -13,7 +13,7 @@
         "WINDOWS_DDPROCMON_VERSION": "1.2.0",
         "WINDOWS_DDPROCMON_SHASUM": "e5525959521254ba8218f64f644ad19cdeb720aa48ee83eb5e402a8d1643e32e",
-        "INTEGRATIONS_CORE_VERSION": "700e8b093d891d2a4b45ab653bfa9a1d16f2bea2"
+        "INTEGRATIONS_CORE_VERSION": "regis.desgroppes/macos-aarch64"
```
to be removed

rdesgroppes replied:
Done - resolving.

git-thuerk-done reviewed (approved):
good for docs

alopezz commented on .gitlab/functional_test/static_quality_gate.yml:
```diff
@@ -51,6 +51,8 @@ static_quality_gates:
     - windows_msi_and_bosh_zip_x64-a7
  #  - job: agent_dmg-x64-a7
  #    optional: True
+ #  - job: agent_dmg-arm64-a7
+ #    optional: True
```
Not sure whether we should add this here or rather remove the existing commented out bit altogether.

alopezz commented on .gitlab/deploy_packages/nix.yml:
```diff
@@ -146,17 +146,29 @@ deploy_packages_dogstatsd_suse_rpm-x64-7:
     PACKAGE_ARCH: x86_64
 
-deploy_packages_dmg-x64-a7:
+.deploy_packages_dmg:
```
To be consistent with other deploy job templates, we may want to move this to https://github.com/DataDog/datadog-agent/blob/regis.desgroppes/macos-aarch64/.gitlab/deploy_packages/deploy_common.yml.

alopezz commented on .gitlab/deploy_packages/nix.yml:
```diff
@@ -146,17 +146,29 @@ deploy_packages_dogstatsd_suse_rpm-x64-7:
+
+deploy_packages_dmg-arm64-a7:
+  extends: .deploy_packages_dmg
+  tags: ["arch:arm64"]
```
I think we can use amd64 for these jobs like we do with others: https://github.com/DataDog/datadog-agent/blob/31da4bbe17af49740f04ce0023c216808aa4d84d/.gitlab/deploy_packages/deploy_common.yml#L5

Since all this does is really just uploading a package from gitlab to S3 it doesn't really need to match the architecture of the package itself.

alopezz replied:
To clarify what I mean: this is more about moving this line to the template rather than about the specific runner to use, as it's independent of the architecture that the artifact is meant for.

alopezz commented on pkg/process/checks/host_info_test.go:
```diff
@@ -100,7 +100,7 @@ func TestGetHostnameFromCmd(t *testing.T) {
 
 func TestResolveHostname(t *testing.T) {
-	if os.Getenv("CI") == "true" && runtime.GOOS == "darwin" && runtime.GOARCH == "amd64" {
+	if os.Getenv("CI") == "true" && runtime.GOOS == "darwin" && (runtime.GOARCH == "amd64" || runtime.GOARCH == "arm64") {
```
I think we may as well drop the part about the architecture, I can't think of a good reason to keep it:
```suggestion
	if os.Getenv("CI") == "true" && runtime.GOOS == "darwin" {
```

Same for the rest of the occurrences (I won't add extra comments for those).

rdesgroppes committed:
Remove commented out jobs

This is to address:
https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2192238875

rdesgroppes committed:
Move DMG deploy template to `deploy_common.yml`

This is to address: (both)
- https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2192246476
- https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2192250329

rdesgroppes committed:
Extend skipped tests for `darwin` to all its archs

This is to address:
https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2192259372

chouetz reviewed (approved):
I don't know if this is a valid question, but we don't have any impact with respect to the go build tags?
And it there anything to change in the macos install script as well? 
Thanks in advance

CelianR commented on tasks/omnibus.py:
```diff
@@ -110,7 +110,7 @@ def get_omnibus_env(
 
     if sys.platform == 'darwin':
         # Target MacOS 10.12
```
```suggestion
```

rdesgroppes committed:
Remove redundant comment

This is to address:
https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2219484524

rdesgroppes committed:
Remove useless `-7` suffix

This is to address:
https://github.com/DataDog/datadog-agent/pull/37676#discussion_r2219375760

rdesgroppes commented on .gitlab/deploy_packages/deploy_common.yml:
```diff
@@ -24,6 +24,17 @@
   variables:
     MAJOR_VERSION: 7
 
+.deploy_packages_dmg-7:
```
:eyes: 17ddcf6160f097592dce6ad64fca5f0282a72c8d

alopezz reviewed (approved):
Nice work! 👍

rdesgroppes commented on .gitlab/source_test/macos.yml:
```diff
@@ -35,19 +40,7 @@ include:
 tests_macos_gitlab_arm64:
   extends: .tests_macos_gitlab
-  rules:
-    !reference [.manual]
```
Before, theses tests had to be triggered manually.
